### PR TITLE
Multi-module single app

### PR DESCRIPTION
**Pull request by ipetr0v**
This commit:
- Makes `app_config_serializer` to use `textproto` files instead of flags with parameters
- Adds the ability to serialize applications containing multiple Wasm modules

### Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [x] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: scripts/build_example [hunk 1]
```diff
@@ -55,7 +55,9 @@ bazel_build_flags+=(
 
 # Use a different output_base so that we don't lose incremental state.
 # See https://docs.bazel.build/versions/master/command-line-reference.html#flag--output_base.
-bazel --output_base="$CACHE_DIR/client" build "${bazel_build_flags[@]}" "//examples/$TARGET/client"
+bazel --output_base="$CACHE_DIR/client" build "${bazel_build_flags[@]}" \
+    "//examples/$TARGET/client" \
+    "//oak/common:app_config_serializer"
```
Comment (daviddrysdale) [inline scripts/build_example]: Why is this added here?

File: oak/common/BUILD [hunk 2]
```diff
@@ -111,6 +111,8 @@ cc_binary(
         "//oak/common:utils",
```
Comment (daviddrysdale) [inline oak/common/BUILD]: Is app_config still needed as a dependency?

File: oak/common/app_config_serializer.cc [hunk 3]
```diff
@@ -14,46 +14,71 @@
...
-    oak::AddStorageToConfig(application_config.get(), storage_address);
+
+  // Load application configuration.
+  auto config = absl::make_unique<oak::ApplicationConfiguration>();
+  std::string textproto_string = oak::utils::read_file(textproto);
+  google::protobuf::TextFormat::MergeFromString(textproto_string, config.get());
+
+  // Add Wasm module bytes to the application configuration.
+  for (auto& node_config : *config->mutable_node_configs()) {
+    auto it = module_map.find(node_config.name());
```
Comment (daviddrysdale) [inline oak/common/app_config_serializer.cc]: Do you need to `continue` if the `node_config` is not for a Wasm Node?

File: oak/common/app_config_serializer.cc [hunk 4]
```diff
@@ -14,46 +14,71 @@
...
+    if (it != module_map.end()) {
+      std::string module_bytes = oak::utils::read_file(it->second);
```
Comment (daviddrysdale) [inline oak/common/app_config_serializer.cc]: Might be helpful to check for a missing file / empty `module_bytes` and error out.

File: oak/common/app_config_serializer.cc [hunk 5]
```diff
@@ -14,46 +14,71 @@
...
+#include "google/protobuf/text_format.h"
 #include "oak/common/app_config.h"
 #include "oak/common/utils.h"
 
-ABSL_FLAG(std::string, module, "", "File containing the compiled WebAssembly module");
-ABSL_FLAG(bool, logging, false, "Enable application logging");
-ABSL_FLAG(std::string, storage_address, "127.0.0.1:7867",
-          "Address ot the storage provider to connect to");
-ABSL_FLAG(int, grpc_port, 8080, "Port for the Application to listen on");
+ABSL_FLAG(std::string, textproto, "", "Textproto file with application configuration");
```
Comment (daviddrysdale) [inline oak/common/app_config_serializer.cc]: Mention that any module_bytes content in the existing config should be empty / will be overwritten?

File: oak/common/app_config_serializer.cc [hunk 6]
```diff
@@ -14,46 +14,71 @@
...
+ABSL_FLAG(std::vector<std::string>, modules, std::vector<std::string>{},
+          "A comma-separated list of entries `module=path` "
```
Comment (daviddrysdale) [inline oak/common/app_config_serializer.cc]: 'module' is supposed to match the name of a `NodeConfiguration`, yes?  It would be nice to find a better label but I can't immediately think of one.

File: oak/common/app_config_serializer.cc [hunk 7]
```diff
@@ -14,46 +14,71 @@
...
+          "with files containing compiled WebAssembly modules");
```
Comment (daviddrysdale) [inline oak/common/app_config_serializer.cc]: Maybe "...to insert into the generated configuration"

File: scripts/build_example (hunk 1 above)
Comment (ipetr0v) [reply scripts/build_example]: It will be used at some point to create configuration files for examples.'
Deleted.

File: oak/common/app_config_serializer.cc (hunk 4 above)
Comment (ipetr0v) [reply oak/common/app_config_serializer.cc]: Done.
Missing file property is checked in `oak::utils::read_file`:
https://github.com/project-oak/oak/blob/e6e6b5532017f11d3f624a279e4e2e80ca14959b/oak/common/utils.cc#L27-L29

File: oak/common/app_config_serializer.cc (hunk 3 above)
Comment (ipetr0v) [reply oak/common/app_config_serializer.cc]: Done

File: oak/common/app_config_serializer.cc (hunk 7 above)
Comment (ipetr0v) [reply oak/common/app_config_serializer.cc]: Done

File: oak/common/app_config_serializer.cc (hunk 6 above)
Comment (ipetr0v) [reply oak/common/app_config_serializer.cc]: Yes.
We need to somehow identify them, I thought it was the easiest option.

File: oak/common/app_config_serializer.cc (hunk 5 above)
Comment (ipetr0v) [reply oak/common/app_config_serializer.cc]: Done

File: oak/common/BUILD (hunk 2 above)
Comment (ipetr0v) [reply oak/common/BUILD]: Yes, it uses `oak::utils::read_file`

File: oak/common/app_config_serializer.cc [hunk 8]
```diff
@@ -14,46 +14,80 @@
...
 #include "oak/common/utils.h"
 
-ABSL_FLAG(std::string, module, "", "File containing the compiled WebAssembly module");
-ABSL_FLAG(bool, logging, false, "Enable application logging");
-ABSL_FLAG(std::string, storage_address, "127.0.0.1:7867",
-          "Address ot the storage provider to connect to");
-ABSL_FLAG(int, grpc_port, 8080, "Port for the Application to listen on");
+ABSL_FLAG(std::string, textproto, "",
+          "Textproto file with application configuration, where the `module_bytes` value is empty, "
+          "(it will be overritten by module bytes after serialization)");
```
Comment (daviddrysdale) [inline oak/common/app_config_serializer.cc]: nit: s/overritten/overwritten/

File: oak/common/BUILD (hunk 2 above)
Comment (daviddrysdale) [reply oak/common/BUILD]: Sorry, GitHub only let me comment on the wrong line: I meant `//oak/common:app_config` on the line above.

File: oak/common/app_config_serializer.cc (hunk 8 above)
Comment (ipetr0v) [reply oak/common/app_config_serializer.cc]: Done

File: oak/common/BUILD (hunk 2 above)
Comment (ipetr0v) [reply oak/common/BUILD]: It's needed, since the serializer uses `oak::WriteConfigToFile`:
https://github.com/project-oak/oak/blob/e6e6b5532017f11d3f624a279e4e2e80ca14959b/oak/common/app_config.h#L33-L34
